controllers/cps_controller_helper.go

```diff
@@ -0,0 +1,217 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	cacheRuntime "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+var setupLog = ctrl.Log.WithName("setup")
+
+func (r *GatekeeperReconciler) addConstraintController(mainCtx context.Context,
+	gatekeeper *operatorv1alpha1.Gatekeeper,
+) error {
+	ready, err := checkCPScontrollerAvailable(mainCtx, gatekeeper, r.DynamicClient)
+	if err != nil {
+		return err
+	}
+
+	// auditFromCache is not Automatic, Delete existing ConstraintPodStatus Controller
+	if !ready {
+		if r.childCtx != nil {
+			r.DeleteCPSController()
+		}
+
+		return nil
+	}
+
+	if r.isCpsCtlInstalled {
+		return nil
+	}
+
+	r.childCtx, r.childCtxCancel = context.WithCancel(mainCtx)
+
+	cpsMgr, err := ctrl.NewManager(r.Cfg, ctrl.Options{
+		Scheme: r.Scheme,
+		Metrics: server.Options{
+			BindAddress: ":8083",
+		},
+		HealthProbeBindAddress: ":8084",
+		LeaderElection:         r.EnableLeaderElection,
+		LeaderElectionID:       "5ff985ccc.constraintstatuspod.gatekeeper.sh",
+		Cache: cacheRuntime.Options{
+			ByObject: map[client.Object]cacheRuntime.ByObject{
+				&gkv1beta1.ConstraintPodStatus{}: {
+					Transform: func(obj interface{}) (interface{}, error) {
+						constraintStatus := obj.(*gkv1beta1.ConstraintPodStatus)
+						// Only cache fields that are utilized by the controllers.
+						guttedObj := &gkv1beta1.ConstraintPodStatus{
+							TypeMeta: constraintStatus.TypeMeta,
+							ObjectMeta: metav1.ObjectMeta{
+								Name:      constraintStatus.Name,
+								Labels:    constraintStatus.Labels,
+								Namespace: constraintStatus.Namespace,
+							},
+							Status: gkv1beta1.ConstraintPodStatusStatus{
+								ObservedGeneration: constraintStatus.Status.ObservedGeneration,
+								Operations:         constraintStatus.Status.Operations,
+							},
+						}
+
+						return guttedObj, nil
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		setupLog.Error(err, "Failed to setup NewManager for ConstraintPodstatus contoller")
+
+		return err
+	}
+
+	// Use another go routine for gatekeeper controller
+	go func() {
+		err = r.addConstraintStatusManager(cpsMgr)
+		if err != nil {
+			r.isCpsCtlInstalled = false
+
+			setupLog.Error(err, "unable to start ConstraintPodStatus manager")
+		}
+	}()
+
+	return nil
+}
+
+func (r *GatekeeperReconciler) addConstraintStatusManager(mgr manager.Manager,
+) error {
+	if err := (&ConstraintStatusReconciler{
+		Scheme:               r.Scheme,
+		Client:               mgr.GetClient(),
+		ClientSet:            kubernetes.NewForConfigOrDie(mgr.GetConfig()),
+		DynamicClient:        r.DynamicClient,
+		Log:                  ctrl.Log.WithName("controllers").WithName("ConstraintPodStatus"),
```

Comment:
Is this what you meant?
```suggestion
		Log:                  ctrl.Log.WithName("ConstraintPodStatus"),
```